oprf/oprf.go

```diff
@@ -0,0 +1,222 @@
+	"errors"
+	"fmt"
+
+	"github.com/cloudflare/circl/oprf/group"
+)
+
+const (
+	// OPRFP256 is the constant to represent the OPRF P-256 with SHA-512 (SSWU-RO) group.
+	OPRFP256 uint16 = 0x0003
+	// OPRFP384 is the constant to represent the OPRF P-384 with SHA-512 (SSWU-RO) group.
+	OPRFP384 uint16 = 0x0004
+	// OPRFP521 is the constant to represent the OPRF P-521 with SHA-512 (SSWU-RO) group.
+	OPRFP521 uint16 = 0x0005
+)
+
+const (
+	// OPRFMode is the context string to define a OPRF.
+	OPRFMode string = "000"
+)
+
+var (
+	// ErrUnsupportedGroup is an error stating that the ciphersuite chosen is not supported
+	ErrUnsupportedGroup = errors.New("the chosen group is not supported")
+)
+
+// BlindToken corresponds to a token that has been blinded.
+// Internally, it is a serialized Element.
+type BlindToken []byte
+
+// IssuedToken corresponds to a token that has been issued.
+// Internally, it is a serialized Element.
+type IssuedToken []byte
+
+// Token is the object issuance of the protocol.
+type Token struct {
+	data  []byte
+	blind *group.Scalar
+}
+
+// Evaluation corresponds to the evaluation over a token.
+type Evaluation struct {
+	element []byte
+}
+
+// KeyPair is an struct containing a public and private key.
+type KeyPair struct {
+	PubK  *group.Element
+	PrivK *group.Scalar
+}
+
+// Client is a representation of a Client during protocol execution.
+type Client struct {
+	suite *group.Ciphersuite
+	ctx   string
+}
+
+// ClientContext implements the functionality of a Client.
+type ClientContext interface {
+	// Blind generates a token and blinded data.
+	Blind(in []byte) (*Token, *BlindToken, error)
+
+	// Unblind unblinds the server response.
+	Unblind(t *Token, e *Evaluation) (IssuedToken, error)
+
+	// Finalize outputs a byte array that corresponds to its input.
+	Finalize(t *Token, issuedT IssuedToken, info []byte) []byte
+}
+
+// Server is a representation of a Server during protocol execution.
+type Server struct {
+	suite *group.Ciphersuite
+	ctx   string
+	Keys  *KeyPair
+}
+
+// ServerContext implements the functionality of a Server.
+type ServerContext interface {
+	// Evaluate evaluates the token.
+	Evaluate(b BlindToken) *Evaluation
+}
+
+func generateCtx(suiteID uint16) string {
+	return OPRFMode + fmt.Sprintf("%x", suiteID)
+}
+
+func generateKeyPair(suite *group.Ciphersuite) *KeyPair {
+	privK := suite.RandomScalar()
+	pubK := suite.ScalarMultBase(privK)
+
+	return &KeyPair{pubK, privK}
+}
+
+func suiteFromID(suiteID uint16) (*group.Ciphersuite, error) {
+	var err error
+	var suite *group.Ciphersuite
+
+	switch suiteID {
+	case OPRFP256:
+		suite, err = group.NewSuite("P-256")
+	case OPRFP384:
+		suite, err = group.NewSuite("P-384")
+	case OPRFP521:
+		suite, err = group.NewSuite("P-521")
+	default:
+		return suite, ErrUnsupportedGroup
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	return suite, nil
+}
+
+// NewServer creates a new instantiation of a Server.
+func NewServer(suiteID uint16) (*Server, error) {
```

Comment:
@tomrist see here.

Reply:
integer IDs for ciphersuites sounds much better. Are these constant values defined in RFC?

Reply:
Yes they are. They are on the suites section.